pkg/actors/targets/workflow/orchestrator/activity.go

```diff
@@ -67,15 +67,29 @@ func (o *orchestrator) callActivity(ctx context.Context, e *backend.HistoryEvent
 		return err
 	}
 
-	targetActorID := buildActivityActorID(o.actorID, e.GetEventId(), generation)
+	activityActorType := o.activityActorType
+	var targetAppID, sourceAppID string
+
+	if router := e.GetRouter(); router != nil {
+		targetAppID = router.GetTarget()
+		sourceAppID = router.GetSource()
+		log.Debugf("Cross-app activity call: target appID=%s, source appID=%s", targetAppID, sourceAppID)
+	}
+
+	// If target app is specified and different from current app, use cross-app actor type
+	if targetAppID != "" && targetAppID != o.appID {
+		activityActorType = fmt.Sprintf("dapr.internal.%s.%s.activity", o.namespace, targetAppID)
+	}
```

Comment:
Shouldn't check empty string for existence.

Use helper func for string construction.